Tests/scripts/lock_cloud_machines.py

```diff
@@ -172,81 +201,158 @@ def try_to_lock_machine(storage_bucket: any, machine: str, machines_locks: list,
+                                                gcs_locks_path, MACHINES_LOCKS_REPO)
+
+    lock_machine_list = []
+    while number_machines_to_lock > 0:
+        busy_machines = []
+        for machine in list_machines:
+            lock_machine_name = try_to_lock_machine(storage_bucket, machine, machines_locks,
+                                                    gitlab_status_token, gcs_locks_path,
+                                                    job_id)
+
+            # We managed to lock a machine
+            if lock_machine_name:
+                lock_machine_list.append(lock_machine_name)
+                number_machines_to_lock -= 1
+                # If we don't need more machines to lock we end the loop
+                if not number_machines_to_lock:
+                    break
+
+                # If the machine was busy we save it to try it again later
+            else:
+                busy_machines.append(machine)
+
+        # Next round we will try and lock only the busy machines
+        list_machines = busy_machines
+
+        # we need more machines but all machines where busy in this round
+        if number_machines_to_lock:
+            sleep(60)
+    return lock_machine_list
+
+
+def create_list_of_machines_to_run(storage_bucket, lock_machine_name, gcs_locks_path, test_machines_path,
+                                   number_machines_to_lock):
+    """
+    get the list of the available machines (or one specific given machine for debugging).
+    Args:
+        storage_bucket(google.cloud.storage.bucket.Bucket): google storage bucket where lock machine is stored.
+        lock_machine_name(str): the name of the file with the list of the test machines.
+        test_machines_path(str): all the exiting machines.
+        gcs_locks_path(str): the lock repository name.
+        number_machines_to_lock(int): the number of the requested machines.
+
+    Returns: the machines list.
+    """
+    if lock_machine_name:  # For debugging: We got a name of a specific machine to use.
+        logging.info('trying to lock the given machine: {options.lock_machine_name}')
+        list_machines = [lock_machine_name]
+    else:
+        logging.info('getting all machine names')  # We are looking for a free machine in all the available machines.
+        test_machines_list = storage_bucket.blob(f'{gcs_locks_path}/{test_machines_path}')
+        list_machines = test_machines_list.download_as_string().decode("utf-8").split()
+        random.shuffle(list_machines)
+
+    if number_machines_to_lock > len(list_machines):
+        logging.error(
+            f'This build requested {number_machines_to_lock} but there are only {len(list_machines)} of active machines')
+        raise
+    return list_machines
+
+
+def wait_for_build_to_be_first_in_queue(storage_client, storage_bucket, gcs_locks_path, job_id, gitlab_status_token):
+    """
+    Run until the job-id is the first in the queue.
```

Comment:
As in the 'get_and_lock_all_needed_machines' function let's elaborate a bit more, something like - this function will wait (busy waiting) for the current build to be first, it will check for the predecessor build for pulse (see the build before it is alive) and cancel it in case it is not, between runs it will sleep for a random time of bla bla...